src/dune_engine/action_builder.ml

```diff
@@ -5,67 +5,100 @@ type 'a eval_mode = 'a Rule.eval_mode =
   | Lazy : unit eval_mode
   | Eager : Dep.Fact.t eval_mode
 
+let union : type a. a eval_mode -> a Dep.Map.t -> a Dep.Map.t -> a Dep.Map.t =
+ fun mode a b ->
+  match mode with
+  | Lazy -> Dep.Set.union a b
+  | Eager -> Dep.Facts.union a b
+
+let union_all mode l = List.fold_left l ~init:Dep.Map.empty ~f:(union mode)
+
 module T = struct
-  type 'a t =
-    | Pure : 'a -> 'a t
-    | Map : ('a -> 'b) * 'a t -> 'b t
-    | Bind : 'a t * ('a -> 'b t) -> 'b t
-    | Both : 'a t * 'b t -> ('a * 'b) t
-    | Seq : unit t * 'b t -> 'b t
-    | All : 'a t list -> 'a list t
-    | Map2 : ('a -> 'b -> 'c) * 'a t * 'b t -> 'c t
-    | Paths_glob : File_selector.t -> Path.Set.t t
-    | Source_tree : Path.t -> Path.Set.t t
-    | Dep_on_alias_if_exists : Alias.t -> bool t
-    | If_file_exists : Path.t * 'a t * 'a t -> 'a t
-    | Contents : Path.t -> string t
-    | Lines_of : Path.t -> string list t
-    | Dyn_paths : ('a * Path.Set.t) t -> 'a t
-    | Dyn_deps : ('a * Dep.Set.t) t -> 'a t
-    | Fail : fail -> _ t
-    | Deps : Dep.Set.t -> unit t
-    | Memo_build : 'a Memo.Build.t -> 'a t
-    | Dyn_memo_build : 'a Memo.Build.t t -> 'a t
-    | Goal : 'a t -> 'a t
-    | Push_stack_frame :
-        (unit -> User_message.Style.t Pp.t) * (unit -> 'a t)
-        -> 'a t
-    | Thunk : 'a Rule.thunk -> 'a t
-
-  let return x = Pure x
-
-  let map x ~f = Map (f, x)
-
-  let bind x ~f = Bind (x, f)
-
-  let both x y = Both (x, y)
-
-  let all xs = All xs
-
-  let memo_build f = Memo_build f
+  open Memo.Build.O
+
+  type 'a t = 'a Rule.thunk =
+    { f : 'm. 'm eval_mode -> ('a * 'm Dep.Map.t) Memo.Build.t }
+  [@@unboxed]
+
+  let return x = { f = (fun _mode -> Memo.Build.return (x, Dep.Map.empty)) }
+
+  let map t ~f =
+    { f =
+        (fun mode ->
+          let+ x, deps = t.f mode in
+          (f x, deps))
+    }
+
+  let bind t ~f =
+    { f =
+        (fun mode ->
+          let* x, deps1 = t.f mode in
+          let+ y, deps2 = (f x).f mode in
+          (y, union mode deps1 deps2))
+    }
+
+  let both x y =
+    { f =
+        (fun mode ->
+          let+ x, deps1 = x.f mode
```

Comment:
Shouldn't we use `fork_and_join` here?

Reply:
`and+` is actually implemented using `fork_and_join` (in fiber.ml)

Reply:
Oh, I didn't realize that. Looks good, then!